src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java

```diff
@@ -1547,21 +1556,66 @@ public List<Action> retrieveActions(
         }
     }
 
+    private static LoadingCache<String, Boolean> createPrivateModeCache() {
+        Duration duration = getPrivateModeCacheExpiration();
+        return Caffeine.newBuilder().expireAfterWrite(duration).build(key -> {
+            if (key.equals(GitHubServerConfig.GITHUB_URL)) {
+                return false;
+            }
+            try {
+                GitHub.connectToEnterpriseAnonymously(key).checkApiUrlValidity();
+            } catch (MalformedURLException e) {
+                // URL is bogus so there is never going to be an avatar - or anything else come to
+                // think of it
+                return true;
+            } catch (IOException e) {
+                if (e.getMessage().contains("private mode enabled")) {
+                    return true;
+                }
+            }
+            return false;
+        });
+    }
+
+    /**
+     * Checks the SystemProperty <code>org.jenkinsci.plugins.github_branch_source.GitHubSCMNavigator.PRIVATE_MODE_CACHE_EXP</code>
+     * for a configured duration expression.
+     * The expression should be in the format expected by {@link Duration#parse(CharSequence)}.
+     * If the expression fails to parse the default duration (20 Hours) will be used.
+     *
+     * @return the duration in the system property or the default Duration (20 Hours)
+     * @see Duration#parse(CharSequence)
+     */
+    private static Duration getPrivateModeCacheExpiration() {
```

Comment:
```suggestion
    @SuppressFBWarnings(value = "DCN_NULLPOINTER_EXCEPTION", justification = "Intentional")
    private static Duration getPrivateModeCacheExpiration() {
```
Looks like just one spotbugs warning about catching null pointer exception instead of checking for null. Personally I think this way is fine since you have to catch the DateTime anyways